Reject index-states after last known index-state

jasagredo opened the pull request:
The `index-state` warning expresses an internal mismatch in the existing index-states and the requested one, but this does not need an action from the user. On the other hand, requesting an index-state after the last known index means that the user should perform `cabal update`. This PR demotes the first one to verbose mode only and throws an error on the second one.

# QA notes

When building a project with a timestamp for which no index-state exists, cabal now warns the user and picks the previous index-state.

The index-state warning that said:

```
Warning: Requested index-state 2023-04-24T00:00:09Z is newer than
'hackage.haskell.org'! Falling back to older state (2023-04-23T19:32:13Z).
```

1. Setting the index-state in-between two existing index-states warns in verbose mode

```
❯ cabal build all -v
...
Reading available packages of hackage.haskell.org...
Using historical state as of 2023-05-09T13:27:17Z as explicitly requested (via
command line / project configuration)
There is no index-state for 'hackage.haskell.org' exactly at the requested
timestamp (2023-05-09T13:27:17Z). Falling back to the previous index-state
that exists: 2023-05-09T12:58:00Z
index-state(hackage.haskell.org) = 2023-05-09T12:58:00Z (HEAD =
2023-05-09T12:58:00Z)
```

2. Setting the index-state after the most-recent known index-state errors

```
❯ cabal build all
Error: cabal: Latest known index-state for 'hackage.haskell.org'
(2023-05-09T13:27:18Z) is older than the requested index-state
(2024-04-24T00:00:00Z). 
Run 'cabal update' or set the index-state to a value at or before
2023-05-09T13:27:18Z.
```

3. Setting the index-state before the oldest known index-state warns with normal verbosity

```
Warning: There is no index-state for 'hackage.haskell.org' exactly 
at the requested timestamp (1900-01-01T00:00:00Z). Also, there are 
no index-states before the one requested, so the repository 
'hackage.haskell.org' will be empty.
```

---
Please include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [x] Include [manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) if your PR relates to cabal-install.

Bonus points for added automated tests!

ulysses4ever reviewed (dismissed):
LGTM thank you!

Do you think you could factor out the message so that we don’t have two copies of it in the source?

I think putting a line break before “Suggestion:” would make it prettier: there’ll be nice parallelism with “Warning:”. But that’s subjective of course.

jasagredo commented:
That sounds good but there are some problems I think. From what I can see, not having done `cabal update` will not produce any warnings or errors:
```
❯ grep hackage cabal.project
  , hackage.haskell.org 2024-04-24T00:00:00Z

❯ cabal build mylib
Build profile: -w ghc-9.2.7 -O1
In order, the following will be built (use -v for more details):
```

Notice I set the index state to next year.

However this error is printed when there is not an index state at the specified timestamp so that cabal falls back to the previous one.

I think the case you want to express is not exactly this one. However I agree with looking for a rephrasing, what about:

```
Warning: The project requires index-state for hackage.haskell.org to be 2023-04-24T00:00:09Z,
but it is not available or doesn't exist. 
Suggestion: Run `cabal update` to get the latest existing index-state.
Warning: In the meantime, falling back to use the latest one before the requested timestamp: 2023-04-23T19:32:13Z. 
Suggestion: Use 2023-04-23T19:32:13Z as the index-state instead of the provided one. The build plan will be the same.
```

angerman commented:
I don't think cabal should report anything here except with `-v`, _unelss_ the known indices are _older_ than the request index state.

If the known indices are _newer_ than the requested index state, there should be _no noise at all_ (unless explicitly requested with `-v`).

What's the value of telling users that the index state they choose is newer than the last one in the index, when we know the index has _newer_ states as well?

This just leads users to produce patches to adjust the index states precisely to index state changes. _Only to silence cabal_. I don't see any value in these kinds of changes.

Thusly, if latestStateOfIndex(index) > requestedStateOfIndex(index) do nothing.
If requestedStateOfIndex(index) > latestStateOfIndex(index), tell the user they need to run `cabal update` _and_ abort! This is a fatal error.

if `-v` be chatty about it.

michaelpj commented:
> If requestedStateOfIndex(index) > latestStateOfIndex(index), tell the user they need to run cabal update and abort! This is a fatal error.

I will say that I have occasionally used an index-state in the future (e.g. tomorrow) as a shorthand for "now or very soon", but I'd probably live. Perhaps this is the case that _should_ be a warning: that's something that the user might want to address (but might be fine).

jasagredo commented:
What about this:

- Setting the index to a future timestamp
```
❯ /home/javier/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.2.7/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal build all
Error: cabal: Latest known index-state for 'hackage.haskell.org'
(2023-05-09T13:27:18Z) is older than the requested index-state
(2024-04-24T00:00:00Z). Run 'cabal update' or set the index-state to a value
at or before 2023-05-09T13:27:18Z.
```

- Setting the index to a non existing timestamp but older than the on-disk index-state file (no warning)
```
❯ /home/javier/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.2.7/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal build all
Resolving dependencies...
```

- Same but verbose
```
❯ /home/javier/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.2.7/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal build all -v
...
Reading available packages of hackage.haskell.org...
Using historical state as of 2023-05-09T13:27:17Z as explicitly requested (via
command line / project configuration)
There is no index-state for 'hackage.haskell.org' exactly at the requested
timestamp (2023-05-09T13:27:17Z). Falling back to the previous index-state
that exists: 2023-05-09T12:58:00Z
index-state(hackage.haskell.org) = 2023-05-09T12:58:00Z (HEAD =
2023-05-09T12:58:00Z)
```

Kleidukos commented:
Last nitpick: We should have the 

> run `cabal update`

On a new line.

Otherwise, all good for me! And my thanks again for your interest in improving our UΧ.

jasagredo commented:
I fixed up the last commit and the PR description. However this change now would require a changelog entry at least and maybe some documentation, but I cannot do that now, so I'm leaving the checkboxes un-ticked.

jasagredo commented:
I added a changelog entry. This is ready to merge I think

Kleidukos commented:
All good to me! :)

ulysses4ever reviewed (commented):
This patch evolved from a simple rewording to two changes in behavior: one with making cabal less Ben's and one with a new failure condition. This is a significant change. I think someone else should look into it. @gbaz @fgaz @Mikolaj

andreabedini commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -275,17 +276,12 @@ getSourcePackagesAtIndexState verbosity repoCtxt mb_idxState mb_activeRepos = do
-                                   "Requested index-state " ++ prettyShow ts0
-                                ++ " does not exist in '"++ unRepoName rname ++"'!"
-                                ++ " Falling back to older state ("
-                                ++ prettyShow (isiMaxTime isi) ++ ")."
+                info verbosity $
+                     "There is no index-state for '"
+                  ++ unRepoName rname ++ "' exactly at the requested timestamp ("
+                  ++ prettyShow ts0 ++ "). Falling back to the previous index-state that exists: "
+                  ++ prettyShow (isiMaxTime isi)
```
I think this logic is wrong. It reads: "if the specified index-state is not equal to the latest index-state, then use the latest index-state".

This would trigger for all historical index-states, am I wrong?

andreabedini reviewed (changes requested):
See comments inline. Also we need tests to codify the behaviour in all situation. See `cabal-testsuite/PackageTests/NewUpdate/UpdateIndexState/` for inspiration.

Kleidukos commented:
@jasagredo Do you want my help to fix the conflicts of your PR?

jasagredo commented:
I can try to fix it next week is I get some free time after work, if that is OK. I think from @andreabedini's comments that I need to rethink the approach

Kleidukos commented:
@jasagredo yes of course, just do not hesitate to reach out if you're facing difficulties

jasagredo commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -406,6 +406,16 @@ readRepoIndex verbosity repoCtxt repo idxState =
+      case idxState of
+        IndexStateHead -> return ()
+        IndexStateTime ts0
+          | ts < ts0 -> die' verbosity $
```
You are right, we cannot compare on this. However it is the case that this is used for a warning already. That's why we see that nix says "the package list for ... is 739042 days old"

andreabedini commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -321,37 +323,13 @@ getSourcePackagesAtIndexState verbosity repoCtxt mb_idxState mb_activeRepos = do
         return ()
       IndexStateTime ts0 -> do
         when (isiMaxTime isi /= ts0) $
```
```suggestion
        -- isiMaxTime is the latest timestamp in the filtered view returned by `readRepoIndex` above.
        -- It is always true that isiMaxTime is less or equal to a requested IndexStateTime.
        -- When `isiMaxTime isi /= ts0` (or equivalently `isiMaxTime isi < ts0`) it means that ts0 falls
        -- between two timestamps in the index.
        when (isiMaxTime isi /= ts0) $
```

andreabedini commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -480,12 +462,32 @@ readRepoIndex verbosity repoCtxt repo idxState =
+      let latestTime = isiHeadTime isi
+       in case idxState of
+            IndexStateTime t -> when (t > latestTime) $ case repo of
+              RepoRemote{..} ->
+                die' verbosity $ errRequestedIdxIsNewer repoRemote latestTime t
```
In `getSourcePackagesAtIndexState` we warn the user that index-state is ignored for old-format repositories, so I believe this should be a `pure ()`. I wish types would help :-/

jasagredo replied:
Done

andreabedini commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -480,12 +462,32 @@ readRepoIndex verbosity repoCtxt repo idxState =
-        RepoRemote{..} -> warn verbosity $ errOutdatedPackageList repoRemote dt
-        RepoSecure{..} -> warn verbosity $ errOutdatedPackageList repoRemote dt
-        RepoLocalNoIndex{} -> return ()
+    warnIfIndexIsOld isi = do
+      curTime <- getCurrentTime
+      case timestampToUTCTime (isiHeadTime isi) of
```
This is a change in behaviour. Before it would warn the user if the index file has not been touched for 15 days (comparing `getCurrentTime` with `getModificationTime`). What you suggest warns the user when the index head time (latest known index state) is more than 15 days in the past.

I belive the old behaviour was more correct. Think of using a repo that only gets packages added once a month, you'd get cabal constantly telling you to do `cabal update` and doing `cabal update` won't even help, since you'd get the same `isiHeadTime`.

I don't think this change is relevant to the PR, so maybe you can just revert this bit? and back to
```
when (isRepoRemote repo) $ warnIfIndexIsOld =<< getIndexFileAge repo
```
?

Or is there some interaction I cannot see?

andreabedini reviewed (changes requested):
Thanks @jasagredo for taking the time to update this. I think it looks good now.
I am suggesting to leave a little comment to our future selves and to revert the change around `warnIfIndexIsOld`. Let me know if you agree with this.

jasagredo commented:
I have a bunch of redundant imports not :/ will fix those when I have some free time

jasagredo commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -440,15 +421,19 @@ readRepoIndex
   handleNotFound $ do
-    when (isRepoRemote repo) $ warnIfIndexIsOld =<< getIndexFileAge repo
     -- note that if this step fails due to a bad repo cache, the the procedure can still succeed by reading from the existing cache, which is updated regardless.
     updateRepoIndexCache verbosity (RepoIndex repoCtxt repo)
       `catchIO` (\e -> warn verbosity $ "unable to update the repo index cache -- " ++ displayException e)
```
Surprisingly, the tests fail because this line is triggered and this warning is printed. However this did not change in this PR.

andreabedini replied:
:thinking: I agree this is puzzling. I think cabal has a bug where it doesn't create the repo local dir in time (i.e. `packages/repo-name`). I wonder if this is related. But I agree nothing seems to have changed.

andreabedini reviewed (dismissed):
Looks good to me! Thank you @jasagredo <3

andreabedini commented:
@jasagredo I did a strace and compared it with master. This branch does
```
write(1, "Reading available packages of re"..., 46) = 46
openat(AT_FDCWD, "/home/andrea/code/cabal/cabal-testsuite/PackageTests/Get/T7248/cabal.dist/home/.cabal/packages/repo.invalid/00-index.timestamp", O_RDONLY|O_NOCTTY|O_NONBLOCK) = -1 ENOENT (No such file or directory)
write(1, "Using most recent state (could n"..., 56) = 56
stat("/home/andrea/code/cabal/cabal-testsuite/PackageTests/Get/T7248/cabal.dist/home/.cabal/packages/repo.invalid/00-index.cache", 0x4200506480) = -1 ENOENT (No such file or directory)
write(1, "Updating index cache file /home/"..., 153) = 153
openat(AT_FDCWD, "/home/andrea/code/cabal/cabal-testsuite/PackageTests/Get/T7248/cabal.dist/home/.cabal/packages/repo.invalid/00-index.tar", O_RDONLY|O_NOCTTY|O_NONBLOCK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/andrea/code/cabal/cabal-testsuite/PackageTests/Get/T7248/cabal.dist/home/.cabal/packages/repo.invalid/00-index.cache", O_RDONLY|O_NOCTTY|O_NONBLOCK) = -1 ENOENT (No such file or directory)
write(2, "Warning: The package list for 'r"..., 96) = 96
write(2, "You may need to run 'cabal updat"..., 81) = 81
```

While master does

```
write(1, "Reading available packages of re"..., 46) = 46
openat(AT_FDCWD, "/home/andrea/code/cabal/cabal-testsuite/PackageTests/Get/T7248/cabal.dist/home/.cabal/packages/repo.invalid/00-index.timestamp", O_RDONLY|O_NOCTTY|O_NONBLOCK) = -1 ENOENT (No such file or directory)
write(1, "Using most recent state (could n"..., 56) = 56
stat("/home/andrea/code/cabal/cabal-testsuite/PackageTests/Get/T7248/cabal.dist/home/.cabal/packages/repo.invalid/00-index.tar", 0x4200506490) = -1 ENOENT (No such file or directory)
write(2, "Warning: The package list for 'r"..., 96) = 96
write(2, "You may need to run 'cabal updat"..., 81) = 81
```

As you see there is indeed a difference. In particular `00-index.timestamp` and `00-index.cache` are not accessed at all by master.

andreabedini commented:
I think I understand this now. We removed

```haskell
    when (isRepoRemote repo) $ warnIfIndexIsOld =<< getIndexFileAge repo
```
from just before
```haskell
    updateRepoIndexCache verbosity (RepoIndex repoCtxt repo)
      `catchIO` (\e -> warn verbosity $ "unable to update the repo index cache -- " ++ displayException e)
```
but all this happens inside the `handleNotFound` handler. So if the repo is missing, `getIndexFileAge` would cause an exception and `updateRepoIndexCache` is skipped.

andreabedini commented:
Ok, here's the plan:

The call to `updateRepoIndexCache` is useless, as [this comment](https://github.com/haskell/cabal/blob/94c86abdf1ea4d9e79a28b14ca003f381eadf974/cabal-install/src/Distribution/Client/IndexUtils.hs#L515-L516) says. We just remove it along with its warning. It seems that `readPackageIndexCacheFile` will not throw exceptions if the cache is missing or corrupted so, if we are lucky, that should not interfere with `handleNotFound`.

Also (and this might be for another time) but I notice that when a repo index is missing, we warn the user but keep going as nothing happens. Perhaps this should be an error to? (should we consider doing `cabal update` automatically? :speak_no_evil: )

andreabedini commented:
I removed the call to `updateRepoIndexCache` and the test suite seems to pass. Then I ran some tests to check what happens when index and/or cache files are missing and I got this:

```
$ export CABAL_DIR=$(mktemp -d)
$ cd work/foliage # any package really
$ /home/andrea/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.4.5/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal build
Warning: this is a debug build of cabal-install with assertions enabled.
Config file path source is default config file.
Config file not found: /tmp/tmp.9Mz7WUxSm6/config
Writing default configuration to /tmp/tmp.9Mz7WUxSm6/config
Warning: The package list for 'hackage.haskell.org' does not exist. Run 'cabal
update' to download it.
Resolving dependencies...
Error: cabal: Could not resolve dependencies:
[__0] trying: foliage-0.6.0.0 (user goal)
[__1] unknown package: zlib (dependency of foliage)
[__1] fail (backjumping, conflict set: foliage, zlib)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: foliage, zlib
```

This looks ok, we are giving the right message (although I am tempted to say we should quit if the package list is missing).

But we have bumped the verbosity to info, so I tried again with `-v` (still without having done `cabal update`).

```
$ /home/andrea/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.4.5/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal build -v
Warning: this is a debug build of cabal-install with assertions enabled.
this build was affected by the following (project) config files:
- /home/andrea/work/foliage/cabal.project
Running: /home/andrea/.ghcup/bin/ghc-9.2 --print-global-package-db
Reading available packages of hackage.haskell.org...
Using historical state as of 2023-03-17T03:33:00Z as explicitly requested (via
command line / project configuration)
Warning: The package list for 'hackage.haskell.org' does not exist. Run 'cabal
update' to download it.
There is no index-state for 'hackage.haskell.org' exactly at the requested
timestamp (2023-03-17T03:33:00Z). Falling back to the previous index-state
that exists:
Running: /usr/bin/pkg-config --version
...
```

Notice the missing timestamp after `Falling back to the previous index-state that exists:`. It turns out `Timestamp` includes a "null" value which is [pretty-printed as ""](https://github.com/haskell/cabal/blob/94c86abdf1ea4d9e79a28b14ca003f381eadf974/cabal-install/src/Distribution/Client/IndexUtils/Timestamp.hs#L87-L94) :facepalm: I even [got biten](https://github.com/haskell/cabal/pull/8131) by this before!

Doing `cabal update -v` shows the issue persists.

```
$ /home/andrea/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.4.5/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal update
Warning: this is a debug build of cabal-install with assertions enabled.
Config file path source is default config file.
Config file not found: /tmp/tmp.9Mz7WUxSm6/config
Writing default configuration to /tmp/tmp.9Mz7WUxSm6/config
Downloading the latest package list from hackage.haskell.org
Package list of hackage.haskell.org has been updated.
The index-state is set to .
```

Perhaps it is a good time to refactor that Timestamp type?

jasagredo commented:
So I have done a couple of things:

1. Make the 'obvious' change to the Timestamps to have a `NoTimestamp` alternative that will be used when initializing caches.
2. Remove the call to `updateRepoIndexCache` as Andrea suggested.
3. Make cabal die when a package list is missing as Andreas suggested. If a package list is missing, there is not much one can do (unless this is a ghc-boot-libraries only package, should we check for this?)

Let me know what you think about these changes.

```
❯ /home/javier/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.2.8/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal build --disable-tests
Warning: this is a debug build of cabal-install with assertions enabled.
Config file path source is default config file.
Config file not found: /tmp/tmp.EMcYV41f6m/config
Writing default configuration to /tmp/tmp.EMcYV41f6m/config
Error: cabal: The package list for 'hackage.haskell.org' does not exist. Run
'cabal update' to download it.
```

jasagredo commented:
I however get this error sometimes, after these changes:
```
❯ unset CABAL_DIR

~
❯ cabal install fourmolu-0.12.0.0
Warning: Parsing the index cache failed (Data.Binary.Get.runGet at position
16: Non-matching structured hashes: d81bdd513f41b5d7ee4cd28455adadbe;
expected: f46da61e7afa58a5e8fd1d2b6fb79899). Trying to regenerate the index
cache...
Resolving dependencies...
Build profile: -w ghc-9.2.8 -O1
...
```

andreabedini commented:
> 3\. If a package list is missing, there is not much one can do (unless this is a ghc-boot-libraries only package, should we check for this?)

You are bringing up a interesting problem. A package that only depends on boot packages would  still able to build succesfully so we might need to rethink that choice.

> Warning: Parsing the index cache failed (Data.Binary.Get.runGet at position
> 16: Non-matching structured hashes: d81bdd513f41b5d7ee4cd28455adadbe;
> expected: f46da61e7afa58a5e8fd1d2b6fb79899). Trying to regenerate the index
> cache...

I would expect this to happen because of the change to the `Timestamp` type, which is used in the cache. I will try to replicate.

andreabedini commented:
oh well ... :joy:

```
~
❯ export CABAL_DIR=$(mktemp -d)

~
❯ /home/andrea/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.4.5/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal update
Warning: this is a debug build of cabal-install with assertions enabled.
Config file path source is default config file.
Config file not found: /tmp/tmp.3Z3xSFjWH1/config
Writing default configuration to /tmp/tmp.3Z3xSFjWH1/config
Downloading the latest package list from hackage.haskell.org
Error: cabal: The package list for 'hackage.haskell.org' does not exist. Run
'cabal update' to download it.
```

jasagredo commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -948,6 +948,10 @@ withIndexEntries verbosity index callback _ = do
+-- | Read package data from a repository.
+-- Throws IOException if any arise while accessing the index
+-- (unless the repo is local+no-index) and dies if the cache
+-- is corrupted and cannot be regereated correctly.
```
```suggestion
-- is corrupted and cannot be regenerated correctly.
```

jasagredo commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -1116,6 +1126,11 @@ readIndexCache verbosity index = do
+-- | Read a no-index repository cache from the filesystem
+--
+-- If a corrupted index cache is detected this function regenerates
+-- the index cache and then reattempt to read the index once (and
```
```suggestion
-- the index cache and then reattempts to read the index once (and
```

jasagredo commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -1165,15 +1181,27 @@ writeIndexTimestamp index st =
+-- timestamp you would use to revert to this version.
+--
+-- Note: this is not the same as 'readIndexTimestamp'!
+-- This resolves HEAD to the the index isiHeadTime, i.e.
```
```suggestion
-- This resolves HEAD to the index's 'isiHeadTime', i.e.
```

andreabedini commented:
I realised we don't have the right test infrastructure to test anything that has to do with timestamps in repositories. I am putting something together.

andreabedini commented:
I have written an DSL to create secure repositories suitable for tests. Next step is integrating with the cabal-testsuite.

jasagredo commented:
@andreabedini any news on this front?

andreabedini commented:
@jasagredo Sorry, no; other things took priority. I just pushed the prototype dsl to create secure repositories [in this branch](https://github.com/andreabedini/cabal/tree/testsuite/secure-repo) but it still ask to be integrated with the test suite (and the tests written etc).

@erikd said he's available to give a hand pushing this PR through. Maybe we can leave the tests for later ... :man_facepalming:

andreabedini commented on cabal-install/src/Distribution/Client/IndexUtils.hs:
```diff
@@ -942,10 +944,14 @@ withIndexEntries verbosity index callback _ = do
 
+-- | Read package data from a repository.
+-- Throws IOException if any arise while accessing the index
+-- (unless the repo is local+no-index) and dies if the cache
```
```suggestion
-- (unless the repo is local+no-index which has no idex) and dies if the cache
```

erikd reviewed (approved):
I have checked out this branch and messed with the `cabal-install` built from it. Seems to work as advertised.

andreabedini commented:
@erikd can you reproduce https://github.com/haskell/cabal/pull/8944#issuecomment-1645046680?

erikd commented:
> @erikd can you reproduce [#8944 (comment)](https://github.com/haskell/cabal/pull/8944#issuecomment-1645046680)?

I cannot reproduce that:
```
> export CABAL_DIR=$(mktemp -d)
> cabal build all
> dist-newstyle/build/x86_64-linux/ghc-9.6.2/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal update
Warning: this is a debug build of cabal-install with assertions enabled.
Config file path source is default config file.
Config file not found: /tmp/tmp.e2kAGRHBPb/config
Writing default configuration to /tmp/tmp.e2kAGRHBPb/config
Downloading the latest package list from hackage.haskell.org
Package list of hackage.haskell.org has been updated.
```

andreabedini commented:
@jasagredo @erikd so what should we do with this? I can confirm that I cannot reproduce [#8944 (comment)](https://github.com/haskell/cabal/pull/8944#issuecomment-1645046680) anymore.

I think if we add a simple test that checks that `cabal build` fails with a project that says `index-state: 3000-01-01T00:00:00Z` we can then call it done?

The failure message looks ok to me.

```
❯ /home/andrea/code/cabal/dist-newstyle/build/x86_64-linux/ghc-9.4.7/cabal-install-3.11.0.0/x/cabal/build/cabal/cabal build --dry-run --index-state=3000-01-01T00:00:00Z cabal
Warning: this is a debug build of cabal-install with assertions enabled.
Error: cabal: Latest known index-state for 'hackage.haskell.org'
(2023-10-18T02:11:00Z) is older than the requested index-state
(3000-01-01T00:00:00Z).
Run 'cabal update' or set the index-state to a value at or before
2023-10-18T02:11:00Z.
```

andreabedini commented:
Re-reading the code for hackage-repo-tool, I think the timestamp might come directly from the file modification file? if so we could manage to write some tests using `setFileTimes`.

andreabedini commented:
@jasagredo Let me know if there's anything I can do to help push this through.

jasagredo commented:
I think this should be ready to merge! 

Things that changed with my last push:
- Rebased
- Tweaked the warning message so that if there is no previous index at all, tell that the repo will appear empty:
```
Warning: There is no index-state for 'repository.localhost' exactly at the requested timestamp (2023-01-01T00:00:00Z). Also, there are no index-states before the one requested, so the repository 'repository.localhost' will be empty.
```
- Added a test that checks that:
  - Index-state from future is rejected
  - Index-state from before last one warns that the repo will be empty (and fails on resolving dependencies, so it actually accepted the index-state)

jasagredo commented:
The macOS pipelines are failing because macOS sed doesn't work as Linux sed. Will fix it tomorrow morning or tonight.

andreabedini commented:
Good for me. @Mikolaj @ulysses4ever @Kleidukos are you ok with these tests? In particular cabal-testsuite/PackageTests/NewUpdate/RejectFutureIndexStates/cabal.test.hs has to so some tricks because we don't seem to be able to fix the timestamp in the repo created by hackage-repo-tool (@jasagredo correct me if I am getting this wrong).

jasagredo commented:
More than
>  to so some tricks because we don't seem to be able to fix the timestamp in the repo created by hackage-repo-tool
I can only check certain things because of that reason. So I can only check:

- rejected repo in the future
- accepted index-state from before first existing index-state (with warning)

But I cannot check that an "in-between-two-index-states" index-state will use the oldest one.

ulysses4ever commented:
I have never played too much with hackage-repo-tool, so I'm willing to believe in what you, guys, managed to find

Mikolaj commented:
Thanks for the automated tests. If that's the best we can reasonably do, it's fine. Could we perhaps have manual QA to supplement them, to be run after the PR is merged but before the release? 

https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes

(BTW, @Kleidukos, why is https://github.com/orgs/haskell/projects/3 not mentioned there? Not decided yet if that's the best place for the board?)

jasagredo commented:
I updated the PR description to make it clear what should be checked. Shall I add the `merge me` label then? @Mikolaj

Mikolaj commented:
@jasagredo: yes, please. Or `squash+merge_me`, the default choice, at your discretion.

I hope the `needs-manual-qa` label now somehow gets the testing request to the project board.

jasagredo commented:
Windows QA testing confirmed the expected behaviour.

michaelpj commented:
I forgot that this patch landed and tried to use a future index-state to get something that just got released :( I note that my previous comment was wrong - using HEAD is bad because it's much more permissive. I can pretty safely commit an index-state of "start of tomorrow" and open myself to at worst a little bit of non-determinism between now and then, whereas HEAD is a lot more non-determinism until I fix it

I guess I'll just have to be less lazy and write down the actual current time...

jasagredo commented:
Which is actually good for determinism. I was just exactly setting up a PR to ensure this in our repo as we had this exact discussion yesterday https://github.com/IntersectMBO/ouroboros-consensus/pull/1158

9999years commented:
Hello, this patch is making a bunch of `cabal-install` commands fail which previously did not fail, e.g.:

```
$ cabal exec -v -- ghc --print-libdir
this build was affected by the following (project) config files:
- /nix/store/vsjh8lkhafvp90qhvari5k0qd4bcrv1l-cabal.project
Running: /nix/store/gmgzg1l65xy7v8liqrvac96n1hzlb8ab-ghc-9.6.3-with-packages/bin/ghc --print-global-package-db
Reading available packages of hackage.haskell.org...
Using most recent state (could not read timestamp file)
Error: [Cabal-7160]
The package list for 'hackage.haskell.org' does not exist. Run 'cabal update' to download it.
CallStack (from HasCallStack):
  dieWithException, called at src/Distribution/Client/IndexUtils.hs:464:31 in cabal-install-3.12.1.0-1HzjdJjLHzi5FPN9QfJl2w:Distribution.Client.IndexUtils
```

This is frustrating, because I'm using Nix and already have a package database with all the packages I'll need to install available and ready. Cabal shouldn't need to consult the index at all. Passing `--offline` or `--index-state="@$(date +%s)"` doesn't seem to help either.

jasagredo commented:
@9999years I think the usage of `cabal exec` is doing exactly what it is supposed to do:

https://github.com/haskell/cabal/blob/eacb6d0f1904ce218c4d81ba57b28d625d7a4516/cabal-install/src/Distribution/Client/CmdExec.hs#L119-L138

Namely it has to consult the index-state to resolve any executables, or to resolve the `with-compiler` option:

```
➜ cabal exec -- ghc --version
Resolving dependencies...
The Glorious Glasgow Haskell Compilation System, version 9.8.2

~\aa via λ 9.6.6 took 2s
➜ ghc --version
The Glorious Glasgow Haskell Compilation System, version 9.6.6
```

So I think the issue is that you are calling `cabal exec` with a wrong environment, not that cabal errors if the package list does not exist.

9999years commented:
@jasagredo Is that true? I'm not sure why `cabal exec` would need to consult the index, as long as all the packages in correct versions are installed and active in the `ghc-pkg` database. Am I missing something here?

jasagredo commented:
I am unsure but I think loading the configuration of a project is an all-or-nothing action. So there is no such thing as "check my wanted ghc, and also these other options, also check the paths for build-tool-depends that I declare, but do not check the index".

My understanding is that in order to for example provide the build-tool-depends, cabal needs to resolve a build plan, which requires going through the available packages, which requires reading the index.

ulysses4ever commented:
May be much better for visibility to open a new ticket for discussion. 

It feels to me like what the Nix-based workflow needs is the low-level Setup.hs interface rather than a holistic tool for dependency resolution and building like cabal-install. Case in point, the Haskell infrastructure in `nixpkgs` is built around the Setup.hs interface. Trying to prevent cabal-install from executing random parts of its operations sounds like a wrong approach. Maybe there should be another tool that operates on a level somewhere between the low and the high?..

I agree that It's annoying when something that used to work stops working. But changing anything at all breaks someone's workflow (cf. the corresponding xkcd) and we cannot not change a living project. When doing any changes, what should be prioritized: to support the rather exotic case of Nix or general improvements in cabal operation (like the one this patch implements). I'd argue the latter.

pwm commented:
[6142386198](https://github.com/haskell/cabal/pull/8944/commits/6142386198dbe0400b31245cf9c479d3ef692e63) also breaks `active-repositories: none`, which for years was a simple way to tell cabal @9999years 's use-case. I'm now not sure how to tell cabal that I don't need it to manage the package list.

gbaz commented:
Breaking active-repositories: none is very bad and we should fix this asap!

gbaz commented:
@pwm how are you passing the active-repositories flag? not sure if i can reproduce on HEAD.

```
cabal --active-repositories :none build exe:cabal
Warning: this is a debug build of cabal-install with assertions enabled.
Resolving dependencies...
```

ulysses4ever commented:
@gbaz in order to reproduce, you have to remove `.cache/cabal/packages/hackage.haskell.org` (or it's non-XDG equivalent).

pwm commented:
@gbaz I'm setting `active-repositories: none` in `cabal.project`

gbaz commented:
got it -- easy to reproduce now by setting an empty directory as `CABAL_DIR` in the environment.
